pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +410,118 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
+
 	if exist {
-		mustCacheResponse = true
-		for ipStr := range responseIPs {
-			if _, ok := oldDNSMeta.responseIPs[ipStr]; !ok {
-				addressUpdate = true
-				break
+		// Data related to this domain exists in the cache.
+		// Besides presence of existing IPs, two scenarios : 1) May get New IPs 2) Some old IPs may be absent.
+
+		// check for new IPs and these new IPs need to be added with its new TTL value as received in response.
+		for ipStr, ipMeta := range responseIPs {
+			if _, exist := oldDNSMeta.responseIPs[ipStr]; !exist {
+				ipMetaDataHolder[ipStr] = ipWithTTL{
+					ip:             ipMeta.ip,
+					expirationTime: ipMeta.expirationTime,
+				}
+				minTimeToReQuery = minTime(ipMeta.expirationTime, minTimeToReQuery)
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		for ipStr, ipMeta := range oldDNSMeta.responseIPs {
+			if _, exist := responseIPs[ipStr]; !exist {
+				if ipMeta.expirationTime.Before(currentTime) {
+					// this ip is expired and stale, remove it by not including it.
 					addressUpdate = true
 				} else {
-					// Add the unexpired IP entry to responseIP and update the lowest applicable TTL if needed.
-					responseIPs[oldIPStr] = oldIP
-					if oldDNSMeta.expirationTime.Before(recordTTL) {
-						recordTTL = oldDNSMeta.expirationTime
-					}
+					// It hasn't expired yet, so just retain it with its existing expirationTime.
+					ipMetaDataHolder[ipStr] = ipMeta
+					minTimeToReQuery = minTime(ipMeta.expirationTime, minTimeToReQuery)
+				}
+			} else {
+				// This old IP also exists in current response, so update it with max time between received time and its old cached time.
+				expTime := maxTime(responseIPs[ipStr].expirationTime, oldDNSMeta.responseIPs[ipStr].expirationTime)
+				ipMetaDataHolder[ipStr] = ipWithTTL{
+					ip:             ipMeta.ip,
+					expirationTime: expTime,
 				}
+				minTimeToReQuery = minTime(expTime, minTimeToReQuery)
 			}
 		}
+
 	} else {
+		// First time seeing this domain.
+		// check if this needs to be tracked, by checking its presence in the rules.
+		// If a FQDN policy had been applied then there must be rule records but because it's
+		// not in cache hence its FQDN:SelectorItem mapping may not be present.
 		for selectorItem := range f.selectorItemToRuleIDs {
 			// Only track the FQDN if there is at least one fqdnSelectorItem matching it.
 			if selectorItem.matches(fqdn) {
-				mustCacheResponse, addressUpdate = true, true
 				f.setFQDNMatchSelector(fqdn, selectorItem)
+				for ipStr, ipMeta := range responseIPs {
+					ipMetaDataHolder[ipStr] = ipWithTTL{
+						ip:             ipMeta.ip,
+						expirationTime: ipMeta.expirationTime,
+					}
+					minTimeToReQuery = minTime(ipMeta.expirationTime, minTimeToReQuery)
```

Comment:
Since thees lines appear four times at least, might make sense to make it a helper.

Reply:
Yes, that was a nice improvement.